sp70/gnu/awk320/awk.man

AWK                                                                    AWK


NAME

        awk - pattern scanning and processing language


SYNOPSIS

        awk [-ffile] [-Fstr] [-t] [-l] [program] [var=text] [file] ...


DESCRIPTION

Awk scans each input file for lines that match any of a set of patterns 
specified in the program.  With each pattern in the program there can be an 
associated action that will be performed when a line of a file matches the 
pattern.

The AWK program may be specified as a file with the -f option as:

        -ffilename.ext
        -f filename.ext

in which case the AWK program is read from the named file.  If the file does 
not exist then an error message will be printed. 

The AWK program may also be specified as a single argument as:

        filename.ext
        filename[.awk]

or as a valid AWK program:

        { for (i in ARGV) printf ("%d: %s\n", i, ARGV[i]) }

AWK will first try to open the first argument as a file, it it can't open a 
file, it then adds the extension ".awk" and tries again to open a file, 
finally AWK will attempt to read the argument directly as an AWK program. 

If the filename is a minus sign (-) then the AWK program is read from the 
standard input.  The program may then be terminated with either a ctrl-Z or
a period (.) on a line by itself.  The second method is useful for entering 
an AWK program followed by the data for the program.  If no program file is 
specified then the program is read from standard input.

If the -f option is selected the full path/name/extension must be specified.  
If only the filename is specified AWK will first attempt to open the named 
file, then the file with the extension ".AWK", finally AWK will attempt to 
parse the parameter as a program.  Multiple -f options may be used to get
the program source from many files.

Files are read in order, the file name '-' means standard input.  Each line 
is matched against the pattern portion of every pattern-action statement; 
the associated action is performed for each matched pattern.

If a file name has the form variable=value, program variables may be changed 
before a file is read.  The assignment takes place when the argument would be 
treated as the next file to read.   Any assignments before the first file
take place before the first BEGIN block is executed.  An assignment after 
the last file will occur before any END block unless an exit was performed.

        awk "{ print code, NR, $$0 }" code=10 file1 code=75 file2

If no files are specified the input is read from standard input.

An input line is made up of fields separated by the field separator FS.  
The fields are denoted by $1, $2 ...; $0 denotes the entire line:

        $0 = "now is the time"

        $1 = "now"      $2 = "is"
        $3 = "the"      $4 = "time"

with the default FS (white space).  If the field separator is set to comma (,)
with "-F," on the command line then the fields might be:

        $0 = "a, b,c,, ,"

        $1 = "a"        $2 = " b"       $3 = "c"
        $4 = ""         $5 = " "        $6 = ""

A pattern-action statement has the form:

        pattern { action }

A missing { action } has the same effect as { print $0 }, a missing pattern 
always matches. 

Pattern-Actions are separated by semicolons or newlines.  A statement may be 
continued on the next line by putting a backslash (\) at the end of the line. 

        { words += NF }; END { print words }

A pattern is a test that is performed on each input line.  If the pattern 
matches the input line then the corresponding action is performed.

Patterns come in several forms:

Form            Example         Meaning

BEGIN           BEGIN {N=1}     initialize N before input is read
END             END {print N}   print N after all input is read
function        function x(y)   define a function called x
text match      /stop/          line contains the string "stop"
expression      $1 == 3         first field is the number 3
compound        /x/ && NF > 2   more that two fields and contain "x"
range           NR==10,NR==20   records ten through twenty inclusive

BEGIN and END patterns are special patterns that match before any files are 
read and after all files have been read respectivly.  There may be multiple 
occurances of these patterns and the associated actions are executed in the 
order that they occur.  

If there is only a series of BEGIN blocks in the awk program and no other 
pattern/action blocks except function declarations then no input files are 
read.  If only END blocks are defined then all the files are read and NR will 
be set to the number of records in all the files.

        BEGIN { page = 5 }

A function pattern is never matched and serves to declare a user defined 
function.  You can declare a function with more parameters than are 
passed as arguments so that the extra parameters can act as local 
variables.  

        function show(a, i) { for (i in a) print a[i] }

A regular expression by itself is matched against the input record ($0). That 
is "/abc/" is equivalent to "$0 ~ /abc/". 

Any expression will match if it evaluates to != 0 or !="".  Also any logical 
combination of expressions and regular expressions may be used as a pattern.

        FILENAME != oldname && FILENAME != "skip"

The last special pattern is two patterns separated by a comma.  This pattern 
specifies a range of records that match the pattern.  The pattern starts to 
match when the first pattern matches and stops matching when the second 
pattern matches.  If they both match on the same input record then only that 
record will match the pattern.

        /AUTHOR/,/NOTES/

An action is a sequence of statements that are performed when a pattern 
matches.  

A statement can be one of the following: 

        { STATEMENT_LIST }
        EXPRESSION
        print EXPRESSION-LIST
        printf FORMAT, EXPRESSION_LIST
        if ( EXPRESSION ) STATEMENT [ else STATEMENT ]
        for ( VARIABLE in ARRAY ) STATEMENT
        for ( EXPRESSION; EXPRESSION; EXPRESSION) STATEMENT
        while ( EXPRESSION ) STATEMENT
        do STATEMENT while ( EXPRESSION )
        break
        continue
        next
        delete ARRAY[SUBSCRIPT]
        exit [ EXPRESSION ]
        return [EXPRESSION ]

A STATEMENT_LIST is a list of statements separated by newlines or semicolons.
As with pattern-actions statements may be extended over more than one line 
with backslash (\). 
        {
            print "value:", i, \
                  "number:", j
            i = i + $3; j++
        }

Expressions take on string or numeric values depending on the operators.
There is only one string operator, concatenation, indicated by adjacent 
expressions.  The following are the operators in order of increasing 
precedence:

Operation           Operator      Example     Meaning

assignment          = *= /= %=    x += 2      two is added to x
                    += -= ^=
conditional         ?:            x?y:z       if x then y else z
logical OR          ||            x||y        if (x) 1 else if (y) 1 else 0
logical AND         &&            x&&y        if (x) if (y) 1 else 0 else 0
array membership    in            x in y      if (exists(y[x])) 1 else 0
matching            ~ !~          $1~/x/      if ($1 contains x) 1 else 0
relational          == != >       x==y        if (x equals y) 1 else 0
                    <= >= <
concatenation                     "x" "y"     a new string "xy"
add, subtract       +  -          x+y         sum of x and y
mul, div, mod       * / %         x*y         product of x and y
unary plus minus    + -           -x          negative of x
logical not         !             !x          if (x is 0 or null) 1 else 0
exponentiation      ^             x^y         x to the yth power
inc, dec            ++ --         x++         x then add 1 to x
field               $             $3          the 3rd field
grouping            ()            ($1)++      increment the 1st field

Variables may be scalars, array elements (denoted x[i]) or fields (denoted 
$expression).  Variable names begin with a letter or underscore and may 
contain any number of letters, digits, or underscores.

Variables are initialized to both zero and the null string.  Fields and the 
command line arguments will be both string and numeric if they can be 
completely represented as numbers.  The range for numbers is 1E-306..1E306.

Array subscripts may be any string.  Multi dimensional arrays are simulated in 
AWK by concatenating the individual indexes with the subscript separator 
between them.  So array[1,1] is equivalent to array[1 SUBSEP 1].   Individual
array elements may be removed with the delete statement, and the whole array
erased with an assignment to the bare variable.

        delete a[i]             # delete one element
        a = ""                  # delete all elements

Simply referencing an array element will cause it to be created and 
initialized.  To avoid creating unwanted elements use the in operator.

        if (i in a) print a[i]  # print one element (if it exists)
        for (i in a) print a[i] # print all elements (that exist)

Comparison will be numeric if both operands are numeric otherwise a string 
comparison will be made.  Operands will be coerced to strings if necessary.  
Uninitialized variables will compare as numeric if the other operand is 
numeric or uninitialized.  Eg. 2 > "10" and 2 < 10.

There are a number of built in variables they are:

Variable        Meaning                                         Default

ARGC            number of command line arguments                   -
ARGV            array of command line arguments                    -
FILENAME        name of current input file                         -
FNR             record number in current file                      -
FS              controls the input field separator                " "
NF              number of fields in current record                 -
NR              number of records read so far                      -
OFMT            output format for records                        "%.6g"
OFS             output field separator                            " "
ORS             output record separator                           "\n"
RLENGTH         length of string matched by match function         -
RS              controls input record separator                   "\n"
RSTART          start of string match by match function            -
SUBSEP          subscript separator                              "\034"


ARGC and ARGV are the count and values of the command line arguments. ARGV[0] 
is the full path/name of AWK.EXE, and the rest are all the command line 
arguments except the "-F", "-f" and program arguments which are used by AWK.

The field separator is a string that is interpreted as a regular expression.
A single space has a special meaning and is changed to /[ \t]+/, any leading 
spaces or tabs are removed.  A BEGIN action may be used to set the separator 
or it may be set by using the -F command line option. 

        BEGIN { FS = "," }      sets FS to a single comma
        "-F[ ]"                 sets FS to a single space

The record separator is a string that is either a newline or the null string.
If the record separator RS is set to the null string then multi line records 
may be read.  In this case the record separator is an empty line. Setting RS 
to "\n" will restore the default behavior.

There are a number of built in functions:

Function            Value returned

atan2(y, x)         arctangent of y/x           in the range -pi to pi
cos(x)              cosine of x                 x in radians
exp(x)              exponentiation of x         (e ^ x)
gsub(r, s)          number of substitutions     substitute s for all r in $0
gsub(r, s, t)       number of substitutions     substitute s for all r in t     
index(s)            position of s in $0         0 if not in $0
index(s, t)         position of t in s          0 if not in s
int(x)              integer part of x
length(s)           number of characters in s
log(x)              natural log of x
match(s, r)         position of r in s or 0     sets RSTART and RLENGTH
rand()              random number               0 <= rand < 1
sin(x)              sine of x                   x in radians
split(s, a)         number of fields            split s into a on FS 
split(s, a, fs)     number of fields            split s into a on fs
sprintf(f, e, ...)  formatted string
sqrt(x)             square root of x
sub(r, s)           number of substitutions     substitute s for one r in $0
sub(r, s, t)        number of substitutions     substitute s for one r in t
substr(s, p)        substring of s from p to end
substr(s, p, n)     substring of s from p of length n
system(s)           exit status                 execute command s

The numeric procedure srand(x) sets a new seed for the random number 
generator.  srand() sets the seed from the system time.

The regular expression arguments of sub, gsub, and match may be either regular 
expressions delimited by slashes or any expression.  The expression is coerced 
to a string and the resulting string is converted into a regular expression.  
This coersion and conversion occurs every time the procedure is called so the 
regular expression form will always be faster. 

The print and printf statements come in several forms:

Form                            Meaning

print                           print $0 on standard output
print expression, ...           prints expressions separated by OFS
print(expression, ...)  
printf format, expression, ...
printf(format, expression, ...)
print >"file"                   print $0 on file "file"
print >>"file"                  append $0 to file "file"
printf(format, ...) >"file"
printf(format, ...) >>"file"

close("file")                   close the file "file"

The print statement prints its arguments on the standard output, or the 
specified file, separated by the current output field separator, and 
terminated by the output record separator.  The printf statement formats its 
expression-list according to the format.  The file is only opened once unless 
it is closed between executions of the print statement.  A file than is open 
for output must be closed if it is to be used for input.  The "file" argument 
may any expression that evaluates to a DOS file name. 

There is one function that is used for input.  It has several forms

Form                            Meaning

getline                         read the next record into $0
getline s                       read the next record into s
getline <"file"                 read a record from file "file" into $0
getline s <"file"               read a record from file "file" into s

getline returns -1 if there is an error (such as non existent file), 0 on 
end of file and 1 otherwise.  The pipe form mentioned in the book is not 
implemented in this version.

The for ( i in a ) statement assigns to i the indexes of a for all elements 
in a.  The while (), do while (), and for (;;) statement is as in C as are 
break and continue. 

The next statements stops processing the pattern action statements and reads 
in the next record.  An exit will cause the END actions to be performed or if 
encountered in an END action will cause termination of the program.  The 
optional expression is returned as the exit status unless overridden by a 
further exit statement in an END action. 

The return statement may be used only in function declarations.  It may have 
an option value to return as the value of the function.  The value of a 
function defaults to zero/null (0/""). 

REGULAR EXPRESSIONS

A \ followed by a single character matches that character.

The ^ matches the beginning of the string.

The $ matches the end of the string.

A . matches any character.

A single character with no special meaning matches that character.

A string enclosed in brackets [] matches any single character in that string.  
Ranges of ASCII character codes may be abbreviated as 'a-z0-9'.  A left 
bracket ] may occur only as the first character of the string.  A literal - 
must be placed where it can't be mistaken as a range indicator. If the first 
character is the caret ^ then any character not in the string will match. 

A regular expression followed by * matches a sequence of 0 or more
matches of the regular expression.

A regular expression followed by + matches a sequence of 1 or more
matches of the regular expression.

A regular expression followed by ? matches a sequence of 0 or 1
matches of the regular expression.

Two adjacent (concatenated) regular expressions match a match of the first 
followed by a match of the second. 

Two regular expressions separated by | match either a match for the
first or a match for the second.

A regular expression enclosed in parentheses matches a match for the
regular expression.

The order of precedence of operators at the same parenthesis level is 
[] then *+? then concatenation then |.


PRINTF FORMAT

Any character except % and \ is printed as that character.

A \ followed by up to three octal digits is the ASCII character
represented by that number.

A \ followed by n, t, r, b, f, v, or p is newline, tab, return, backspace, 

sp66/news

Скоро начнется выпуск новой модификации 486dx2-66

Вышел Dbase V for Windows.

Появилась Chicago beta

Ожидается новая OS/2 for Windows

Workplace OS - операционная система хамелеон


sp70/gnu/awk320/src/awk.c
/*
 * awk main programme
 *
 * Copyright (C) 1988, 1989, 1990, 1991 by Rob Duff
 * All rights reserved
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <mem.h>
#include <alloc.h>
#include <string.h>
#include <setjmp.h>
#include <signal.h>

#include "awk.h"

IDENT *ident;                       /* list of all identifiers */

ITEM *nextvar;                      /* pointer to next unused variable */
ITEM vartab[MAXNAME];               /* table to hold variables */

ITEM *stackptr, *stacktop;          /* stack pointer and top limit */
ITEM stackbot[MAXSTACK];            /* the stack */
ITEM fieldtab[MAXFIELD];            /* table to hold fields 0..MAXFIELD-1 */
RULE *rules, *rulep;                /* awk code pointers */
FYLE files[MAXFILE];                /* table to hold files */
FYLE *nextfile;                     /* pointer to unused file */

LINK *beginact;                     /* list of BEGIN actions */
LINK *beginend;                     /* pointer to last BEGIN action */
LINK *endact;                       /* list of END actions */
LINK *endend;                       /* pointer to last END action */

FILE *awkfile;                      /* AWK program input file */

char *yyname;                       /* program file name */
char *awkfre;                       /* RE for field separator */
char *awkname;                      /* data file name */
char *blankre;                      /* RE for " " field separator */

char fmtstr[66];                    /* format string */
char ofmtstr[66] = "\377%.6g";      /* output format string */
char namebuf[66];                   /* internal form for file name */
char code[MAXCODE];                 /* code generation buffer */
char buffer[MAXCODE];               /* string buffer */
char linebuf[MAXLINE] = "";         /* program input buffer */

NAME *varlist;                      /* list of var=text options */
NAME *varlast;                      /* last var in list */
NAME *awklist;                      /* list of program file names */
NAME *awklast;                      /* last name in list */
char *awkstdin = "<stdin>";

FSTR awkfs;                         /* current field separator */
FSTR lineptr;                       /* program line pointer */
FSTR nullstr = "\377";              /* standard null string */
FSTR blankfs = "\377[ \\t\\n]+";    /* " " field separator */
FSTR fieldbuf;                      /* field string storage */

short trace;                        /* trace execution */
short awkline;                      /* include line information */
short builtin;                      /* number of builtin variables */
short status, awkeof;               /* exit status and EOF status */

ITEM  *ac, *av, *rl, *rst;          /* BUILTIN variables */
ITEM  *nr, *fnr, *fs, *rs;          /* BUILTIN */
ITEM  *ofs, *ors, *ofmt;            /* BUILTIN */
ITEM  *nf, *fn, *nul, *subsep;      /* BUILTIN */

jmp_buf nextjmp, exitjmp;           /* next and exit long jump buffers */

void awkfpe();
extern ELEMENT *allell(void);

static void usage(void);            /* error message */
static void awkinit(void);          /* initialize AWK */
static void execute(void);          /* execute an ACTION */
static void addname(char *);        /* add -f program file */
static void addvar(char *);         /* add -v var=text */
static ITEM *buildin(char*, int, int, FSTR);

int main(int argc, char *argv[])
{
    int     found;                  /* found wildcard file */
    char    *arg0;                  /* program name */
    char    *argn;                  /* data file names */
    LINK    *acts;                  /* pointer to next action */

    awkinit();                      /* initialize program */
    arg0 = argv[0];                 /* get program name */
    argc--; argv++;                 /* get next argument */
    while (argc > 0 && argv[0][0] == '-') { /* parse options */
        if (argv[0][1] == 'f') {    /* input file option */
            if (argv[0][2] != '\0') /* is it a separate arg? */
                addname(argv[0]+2); /* no */
            else if (argc > 0) {
                argc--; argv++;     /* yes */
                addname(argv[0]);   /* get next argument */
            }
            else {                  /* oops no more arguments */
                error("no file specified for -f", NULL);
            }
        }
        else if (argv[0][1] == 'v') {    /* var=text option */
            if (argv[0][2] != '\0') /* is it a separate arg? */
                addvar(argv[0]+2); /* no */
            else if (argc > 0) {
                argc--; argv++;     /* yes */
                addvar(argv[0]);   /* get next argument */
            }
            else {                  /* oops no more arguments */
                error("no var=text specified for -v", NULL);
            }
        }
        else if (argv[0][1] == 'F') {
            argv[0][1] = LSTR;      /* field separator */
            fs->sstr = argv[0] + 1; /* convert to internal form */
            awkeof = 1;
            setfs();                /* calculate RE for FS */
            yyinit();
        }
        else if (argv[0][1] == 't') {
            trace++;                /* trace execution */
        }
        else if (argv[0][1] == 't') {
            awkline++;              /* include lines */
        }
        else if (argv[0][1] == '-') {
            argc--; argv++;         /* "--" means end of options */
            break;
        }
        else if (argv[0][1] == '\0') {
            addname(awkstdin);      /* "-" means stdin */
            argc--; argv++;
            break;
        }
        else
            usage();                /* messy option */
        argc--; argv++;
    }
    if (awklist == NULL) {          /* don't have -f option */
        if (argc > 0) {
            if (strpbrk(*argv, " \",/[]|<>+=;") == NULL) {
                strcpy(buffer, *argv);      /* move to larger space */
                awkfile = fopen(buffer ,"r");
                if (awkfile != NULL) {
                    awkeof = 0;
                    yyname = *argv;
                    lineptr = linebuf;
                    *lineptr = '\0';
                }
                else {
                    strcat(buffer, ".awk"); /* try .AWK */
                    awkfile = fopen(buffer, "r");
                    if (awkfile != NULL) {
                        awkeof = 0;
                        yyname = malloc(strlen(buffer)+1);
                        strcpy(yyname, buffer);
                        lineptr = linebuf;
                        *lineptr = '\0';
                    }
                    else {
                        awkeof = 1;         /* no input files */
                        strcpy(linebuf, *argv);
                        strcat(linebuf, "\n");
                        lineptr = linebuf;
                    }
                }
            }
            else {                  /* must be a program */
                awkeof = 1;         /* no input files */
                strcpy(linebuf, *argv);
                strcat(linebuf, "\n");
                lineptr = linebuf;
            }
            argc--; argv++;
        }
        else {
            awkeof = 0;
            addname(awkstdin);      /* no file name for stdin */
        }
    }
    else
        awkeof = 0;

    yyparse();                      /* compile AWK program */
    awkeof = 1;                     /* end of awk program */
    if (awkfile != stdin)           /* close if not stdin */
        fclose(awkfile);
    deparse();                      /* clean up input record */
    setargv(argc, arg0, argv);      /* set ARGV and ARGC */
    if (varlist != NULL)
        getargs(varlist);
    if (setjmp(exitjmp) == 0) {     /* EXIT() point */
        if (beginact != NULL)       /* if we have BEGIN actions */
            if (setjmp(nextjmp) == 0) { /* set NEXT() point */
                acts = beginact;
                while (acts != NULL) {  /* execute BEGIN actions */
                    awkexec(acts->ccode);
                    acts = acts->cnext; /* point to next action */
                }
            }
        found = 0;
        argn = getargv();
        if (rulep != NULL || endact != NULL) do {
            if (argn != NULL) {
                if (*argn == '\0') {
                    argn = getargv();   /* get the next input file */
                    if (argn == NULL)
                        break;          /* stop when no more */
                }
                if (argn[0] == '-' && argn[1] == '\0') {
                    awkfile = stdin;    /* check for standard input */
                    awkname = "";
                    argn = "";
                }
                else {
                    if (awkfind(buffer, argn, 0) == NULL) {
                        if (found > 0) {    /* wildcard file */
                            found = 0;      /* no more match */
                            argn = "";
                            continue;
                        }
                        error("can't find file %s", argn);
                    }
                    awkfile = fopen(buffer ,"r");
                    if (awkfile == NULL)    /* can't open file */
                        error("can't open data file", buffer);
                    awkname = buffer;       /* set FILENAME */
                    found++;
                }
            }
            else {
                awkname = "";               /* no FILENAME for stdin */
                awkfile = stdin;            /* file not specified */
            }
            namebuf[0] = TSTR;              /* mark FILENAME as temporary */
            strcpy(namebuf+1, awkname);
            setfile(namebuf);               /* set FILENAME and FNR */
            files[1].ffyle = awkfile;
            execute();                      /* execute program */
            if (awkfile != stdin)
                fclose(awkfile);            /* close opened file */
        } while (argn != NULL);             /* repeat for next file */
    }
    deparse();                              /* free FIELD temporaries */
    setfile(nullstr);                       /* clear FILENAME and FNR */
    if (setjmp(exitjmp) == 0) {             /* EXIT () point in END */
        if (setjmp(nextjmp) == 0) {         /* some fool may try this */
            for (acts = endact; acts != NULL; acts = acts->cnext)
                awkexec(acts->ccode);       /* execute end actions */
        }
    }
    awkclose();                             /* close all the FILES */
    return status;                          /* return exit code */
}

static void execute()
{
    int     flag;

    while (getrecord() > 0) {               /* read a record (line) */
        rulep = rules;                      /* start at the first pat/act */
        if (setjmp(nextjmp) == 0) {         /* NEXT() point */
            while (rulep != NULL) {
                if (rulep->start == NULL)
                    rulep->seen = 1;        /* no rule means always */
                else if (rulep->seen == 0 && awkexec(rulep->start) != 0)
                    rulep->seen = 1;        /* success means start */
                flag = rulep->seen;
                if (rulep->stop == NULL)
                    rulep->seen = 0;        /* no rule means once */
                else if (rulep->seen != 0 && awkexec(rulep->stop) != 0)
                    rulep->seen = 0;        /* success means stop */
                if (flag != 0)
                    awkexec(rulep->action); /* execute action on start */
                rulep = rulep->next;        /* do next rule */
            }
        }
        deparse();                          /* free FIELD assignments */
    }
}

/*
 * create builtin variable
 */
static ITEM *buildin(char *name, int class, int type, FSTR ptr)
{
    IDENT *vp;

    nextvar->sclass = class;
    nextvar->stype = type;
    nextvar->sstr = ptr;
    nextvar->svalue.dval = 0;

    vp = malloc(sizeof(IDENT));
    vp->vitem = nextvar;
    vp->vname = name;
    vp->vfunc = NULL;
    vp->vnext = ident;
    ident = vp;
    
    builtin++;
    return nextvar++;
}

/*
 * initialize all the structures used by AWK
 */
static void awkinit()
{
    int     i;
    ELEMENT *ep;

    trace = 0;
    status = 0;
    ident = NULL;
    rules = NULL;
    rulep = NULL;
    endact = NULL;
    endend = NULL;
    beginact = NULL;
    beginend = NULL;
    stackptr = stacktop = stackbot + MAXSTACK;

    fieldbuf = farmalloc(MAXLINE+MAXFIELD+MAXFIELD);
    builtin = 0;
    nextvar = vartab;
    nul = buildin("\"\"", ACTUAL, S_NUMBER, nullstr);
    fs = buildin("FS", ACTUAL, S_STRING, "\377 ");
    ofs = buildin("OFS", ACTUAL, S_STRING, "\377 ");
    nf = buildin("NF", ACTUAL, S_SHORT, 0);
    ofmt = buildin("OFMT", ACTUAL, S_STRING, ofmtstr);

    rs = buildin("RS", ACTUAL, S_STRING, "\377\n");
    ors = buildin("ORS", ACTUAL, S_STRING, "\377\n");
    nr = buildin("NR", ACTUAL, S_SHORT, 0);
    fnr = buildin("FNR", ACTUAL, S_SHORT, 0);
    fn = buildin("FILENAME", ACTUAL, S_STRING, namebuf);

    av = buildin("ARGV", ACTUAL, S_STRING, nullstr);
    ac = buildin("ARGC", ACTUAL, S_STRING, nullstr);

    rl = buildin("RLENGTH", ACTUAL, S_STRING, nullstr);
    rst = buildin("RSTART", ACTUAL, S_STRING, nullstr);
    subsep = buildin("SUBSEP", ACTUAL, S_STRING, "\377\034");

    for (i = builtin; i < MAXNAME; i++) {
        ep = allell();                      /* this stuff allows undefined */
        ep->aclass = ACTUAL;                /* variables to be passed to */
        ep->atype = S_NUMBER;               /* functions as arrays */
        ep->astr = nullstr;
        ep->avalue.dval = 0;
        ep->aindex = nullstr;
        ep->anext = NULL;
        vartab[i].sclass = ACTUAL;
        vartab[i].stype = S_ARRAY;
        vartab[i].sstr = nullstr;
        vartab[i].svalue.aptr = ep;
    }
    for (i = 0; i < MAXFIELD; i++) {
        fieldtab[i].sclass = ACTUAL;
        fieldtab[i].stype = S_STRING;
        fieldtab[i].sstr = nullstr;
    }
    nextfile = files + 2;
    files[0].fname = "\377<stdout>";
    files[0].fmode = "w";
    files[0].ffyle = stdout;
    files[1].fname = "\377<stdin>";
    files[1].fmode = "r";
    files[1].ffyle = stdin;
    for (i = 2; i < MAXFILE; i++) {
        files[i].fname = nullstr;
        files[i].fmode = "";
        files[i].ffyle = NULL;
    }
    awkeof = 1;
    yyinit();
    lineptr = blankfs+1;
    blankre = regexp(2);
    *lineptr = '\0';
    yyinit();
    awkfs = blankfs;
    awkfre = blankre;
    awkfile = stdin;
    awkname = NULL;
    awklist = NULL;
    awklast = NULL;
    varlist = NULL;
    varlast = NULL;
    lineptr = NULL;
    yyname = NULL;
    yyline = 0;

    signal(SIGFPE, awkfpe);
}

/*
 * display an error message and leave
 */
void
error(char *fmt, ...)
{
    va_list ap;

    va_start(ap, fmt);
    fprintf(stderr, "awk ERROR");
    if (yyname)
        fprintf(stderr, " %s", yyname);
    if (yyline)
        fprintf(stderr, " line %d", yyline);
    fprintf(stderr, ": ");
    vfprintf(stderr, fmt, ap);
    fprintf(stderr, "\n");
    va_end(ap)
    exit(1);
}

static void addname(char *name)
{
    NAME    *list;

    list = malloc(sizeof(NAME));
    list->name = name;
    list->next = NULL;
    if (awklast == NULL)
        awklist = list;
    else
        awklast->next = list;
    awklast = list;
}

static void addvar(char *name)
{
    NAME    *list;

    list = malloc(sizeof(NAME));
    list->name = name;
    list->next = NULL;
    if (varlast == NULL)
        varlist = list;
    else
        varlast->next = list;
    varlast = list;
}

void awkfpe()
{
    error("floating point exception", NULL);
}

/*
 * print a help message about how to use AWK and exit
 */
static void usage()
{
    fprintf(stderr, "   AWK version 3.20 (22-May-91)\n\n");
    fprintf(stderr, "          Copyright (C) 1990, 91 by Rob Duff\n");
    fprintf(stderr, "          Vancouver BC Canada V5N 1Y9\n\n");
    fprintf(stderr, "usage: awk [-l] [-t] [-f name] [-v var=text] [-Ffs] [--] [prog] [data ... ]\n");
    fprintf(stderr, "       -l      include line number tracing for debugging\n");
    fprintf(stderr, "       -t      will trace execution of the program\n");
    fprintf(stderr, "       -f      specify program file\n");
    fprintf(stderr, "       name    is the name of a file containing an AWK program. You may\n");
    fprintf(stderr, "               specify multiple program files with multiple -f options\n");
    fprintf(stderr, "               A -f name and prog program definition may not be mixed\n");
    fprintf(stderr, "       -v      assign var before any BEGIN blocks\n");
    fprintf(stderr, "       -F      specify a pattern to match for field separation\n");
    fprintf(stderr, "       fs      is the pattern to match for field separation\n");
    fprintf(stderr, "       --      ends the option list.\n\n");
    fprintf(stderr, "       prog    is a quoted program or a file name with optional [.AWK]\n");
    fprintf(stderr, "               you may specify stdin with a dash (-). Not used with -f.\n");
    fprintf(stderr, "       data    is any file name including wildcards.  If file is of the\n");
    fprintf(stderr, "               form var=text then text is assigned to var in the program\n");
    fprintf(stderr, "               instead of reading a file.\n");
    fprintf(stderr, "       If no options are specified then the program and data are both\n");
    fprintf(stderr, "       read from stdin.  You end the program with a line containing\n");
    fprintf(stderr, "       only a single period (.) in the first column.\n\n");
    exit(1);
}

